crates/sui-core/src/authority/consensus_quarantine.rs

```diff
@@ -330,3 +334,647 @@ impl ConsensusCommitOutput {
+                        .remove(digest)
+                        .unwrap_or_else(|| {
+                            panic!(
+                                "transaction {:?} not found in builder_digest_to_checkpoint",
+                                digest
+                            )
+                        }),
+                    seq
+                );
+            }
+
+            batch.insert_batch(
+                &tables.builder_digest_to_checkpoint,
+                contents.iter().map(|tx| (tx.transaction, seq)),
+            )?;
+
+            batch.insert_batch(
+                &tables.builder_checkpoint_summary_v2,
+                [(seq, &builder_summary)],
+            )?;
+
+            let checkpoint_height = builder_summary
+                .checkpoint_height
+                .expect("non-genesis checkpoint must have height");
+            if let Some(highest) = highest_committed_height {
+                assert!(checkpoint_height > highest);
+            }
+
+            highest_committed_height = Some(checkpoint_height);
+        }
+
+        let Some(highest_committed_height) = highest_committed_height else {
+            return Ok(());
+        };
+
+        while !self.output_queue.is_empty() {
+            // A consensus commit can have more than one pending checkpoint (a regular one and a randomnes one).
+            // We can only write the consensus commit if the highest pending checkpoint associated with it has
+            // been processed by the builder.
+            let Some(highest_in_commit) = self
+                .output_queue
+                .front()
+                .unwrap()
+                .get_highest_pending_checkpoint_height()
+            else {
+                // if highest is none, we have already written the pending checkpoint for the final epoch,
+                // so there is no more data that needs to be committed.
+                break;
+            };
+
+            if highest_in_commit <= highest_committed_height {
+                info!(
+                    "committing output with highest pending checkpoint height {:?}",
+                    highest_in_commit
+                );
+                let output = self.output_queue.pop_front().unwrap();
+                self.remove_shared_object_next_versions(&output);
+                self.remove_processed_consensus_messages(&output);
+                self.remove_congestion_control_debts(&output);
+                epoch_store.remove_shared_version_assignments(
+                    output
+                        .pending_checkpoints
+                        .iter()
+                        .flat_map(|c| c.roots().iter()),
+                );
+                output.write_to_batch(epoch_store, batch)?;
+            } else {
+                break;
+            }
+        }
+
+        Ok(())
+    }
+}
+
+// Read methods - all methods in this block return data from the quarantine which would otherwise
```

Comment:
I see some mutations in this block, not just reads?

Reply:
fixed!